Replace RuntimeErrors with custom exceptions

t00sa opened the pull request:
Swap all RuntimeError exceptions for custom fab exceptions using a class hierarchy derived from RuntimeError to minimise disruptive changes.

These changes fall into three broad areas:

* creation of a new set of error classes in `fab.errors` and an associated set of tests
* changes to the existing regression tests to:
  * check the new message formats
  * check the exceptions - which are still caught as `RuntimeError` to test the impact on calling functions - match the expected instance of the new error class
* replacement of `raise RuntimeError` across the code base with the new error classes

While the change is complete and the test suite runs successfully in my environment, I still have a couple of outstanding questions that need to be addressed in review:

* are the names of the new exceptions classes acceptable?  The top level class names end with `Error` but the derived classes do not.  Ending every class name with `Error` would make fab more consistent with python, at the cost of longer exception names
* are there any potential clashes or issues with the in-flight baf PR?

MatthewHambley commented:
I'm about to start a review but I'll leave this comment here for immediate consideration. Are you aware of `FabException` in https://github.com/MetOffice/fab/blob/main/source/fab/__init__.py? Have you considered how this interacts with your change?

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+
+    def __init__(
+        self,
+        tool: Union["Category", "Tool", str],
+        category: Union["Category", "Tool", type, str],
```
Why can argument `Tool` have a value of type `Category` and argument `category` have value of type `Tool`? Also, why allow strings?

t00sa replied:
I _think_ this is because the `RuntimeError` for which this is a drop-in replacement uses a mixture of tools, categories, and strings in a not particularly intuitive way.  I don't have a strong opinion on whether to leave as is or whether to implement some additional error classes for each pattern.

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+        self,
+        tool: Union["Category", "Tool", str],
+        category: Union["Category", "Tool", type, str],
+        expected: str,
```
Isn't this an expected `Category`?

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+
+    def __init__(
+        self,
+        tool: Union["Category", "Tool", str],
+        value: str,
+        expected: Optional[str] = None,
```
Should `tool` have type `Tool`? is `expected` a regular expression?

t00sa replied:
Tool is also always provided as a string, so I'll update the type hinting accordingly.

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+class FabToolPsycloneAPI(FabToolError):
+    """PSyclone API and target problem.
+
+    Error when the specified PSyclone API, which can be empty to
+    indicate DSL mode, and the associated target do not match.
```
I know this is how PSyclone works but we might prefer to be explicit about the API.

t00sa replied:
I agree: it's not particularly elegant, but I wasn't sure whether it was worth adding more plumbing to raise different sorts of exceptions vs having a generic exception that handles everything

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+    """
+
+    def __init__(
+        self, tool: Union["Category", "Tool", str], suite: Optional[str] = None
```
`tool` of type `Tool`?

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+    :param additional: optional additional information
+    """
+
+    # FIXME: improve these here and in tool_repository
```
Ideally we will eventually remove the need for this failure mode in that the API will not allow impossible requests to be made. Clearly not an issue for this change.

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+    def __init__(
+        self,
+        setting_type: str,
+        tool: Union["Category", "Tool", str],
```
`tool` has type `Tool`?

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+
+    def __init__(
+        self,
+        command: str,
```
The subprocess module also allows commands to be specified as a list of strings. Check how it is done when this exception is raised - we might prefer to accept it straight rather than having to convert it.

In fact, looking below, I see you cover this eventuality so reflect it in the type hinting.

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+        if isinstance(command, list):
+            self.command: str = " ".join(command)
+        else:
+            self.command = str(command)
+        self.code = int(code)
```
Both these things are being cast to the type they are already. Is there a reason for that?

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+        self.code = int(code)
+        self.output = self._decode(output)
+        self.error = self._decode(error)
+        self.cwd = cwd
```
Two types are accepted but we should store just one. I suggest `Path`.

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+    """
+
+    def __init__(self, command: str) -> None:
+        self.command = command
```
Homogenise argument to either string or list of strings. Latter would remove the need for the test to extract target.

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+    multiprocessing section into a single exception class for
+    subsequent inspection.
+
+    This feature is is required because versions of python prior to
```
```suggestion
    This feature is required because versions of python prior to
```

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+    multiprocessing section into a single exception class for
+    subsequent inspection.
+
+    This feature is is required because versions of python prior to
+    3.11 do not support ExceptionGroups.
```
Maybe mark this as `Todo:` so someone can be prompted to remove it once 3.10 is no longer supported.

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+    """
+
+    def __init__(
+        self, errors: List[Union[str, Exception]], label: Optional[str] = None
```
If this exception consolidates exceptions, why is it accepting strings?

If that is useful, should it create an exception to hold the string so `self.errors` holds only exceptions?

MatthewHambley commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+        targeted
+    """
+
+    def __init__(self, tool: str, reason: str, revision: Optional[str] = None) -> None:
```
`tool` has type `Tool`?

MatthewHambley commented on source/fab/steps/grab/prebuild.py:
```diff
@@ -28,4 +29,4 @@ def grab_pre_build(config, path, allow_fail=False):
         msg = f"could not grab pre-build '{path}':\n{err}"
         logger.warning(msg)
```
Do we want to log a warning *and* raise an exception?

MatthewHambley commented on tests/unit_tests/test_cui_arguments.py:
```diff
@@ -101,7 +101,7 @@ def test_nonexistent_user_file(self, fs: FakeFilesystem, capsys):
 
         captured = capsys.readouterr()
-        assert "error: fab file does not exist" in captured.err
+        assert "fab file does not exist" in captured.err
```
Where we are testing error messages I think we should test as much as possible. i.e. prefer equality over "in".

MatthewHambley commented on tests/unit_tests/test_errors.py:
```diff
@@ -0,0 +1,179 @@
+        err = FabToolError("cc", "compiler message")
+        assert str(err) == "[cc] compiler message"
+
+        # Mock defines an internal name property, so special handling
+        # is required to reset the value to support testing
+        tool = Mock(_name="tool cc")
+        del tool.name
+        err = FabToolError(tool, "compiler message")
+        assert str(err) == "[tool cc] compiler message"
```
This seems alarming. Is this related to Joerg's comment about peeking at private state?

MatthewHambley commented on tests/unit_tests/test_errors.py:
```diff
@@ -0,0 +1,179 @@
+        """Test FabCommandError in various configurations."""
+
+        err = FabCommandError(["ls", "-l", "/nosuch"], 1, b"", b"ls: cannot", "/")
+        assert str(err) == "command 'ls -l /nosuch' returned 1"
```
Would it be easier to read if the sentence was restructured to put the variable length element last. e.g. "Return code 1 from command: ls -l /nosuch".

This way the eye knows where to go to for the error code and the start of the command.

MatthewHambley commented on tests/unit_tests/test_errors.py:
```diff
@@ -0,0 +1,179 @@
+        assert str(err) == "[git] merge failed: conflicting source files"
+
+        err = FabSourceMergeError("git", "conflicting source files", "vn1.1")
+        assert str(err) == "[git] merge of vn1.1 failed: conflicting source files"
```
Maybe quote marks around revision?

MatthewHambley commented on tests/unit_tests/tools/test_linker.py:
```diff
@@ -124,7 +124,7 @@ def test_get_lib_flags_unknown(stub_c_compiler: CCompiler) -> None:
     with raises(RuntimeError) as err:
         linker.get_lib_flags("unknown")
-    assert str(err.value) == "Unknown library name: 'unknown'"
+    assert str(err.value) == "unknown library unknown"
```
This might benefit from a redrafting. Initial upper-case and maybe quote marks around library name.

MatthewHambley reviewed (changes requested):
This is an extensive change which tidies up exception handling considerable. A few questions and suggestions included in-line.

In general, we have generated API documentation but is there anything which could be added (or removed) from the static documentation in light of this change?

t00sa commented on source/fab/errors.py:
```diff
@@ -0,0 +1,344 @@
+
+if TYPE_CHECKING:
+    from fab.tools.category import Category
+    from fab.tools.tool import Tool
```
It's certainly unpleasant.  I _think_ it's what I meant when I mentioned the dependency later on.  I think Joerg has plans to deal with it at some point?

t00sa reviewed (commented):
Clearly, some things I still need to tidy up.  I'll get on with them tomorrow and update the PR when they're ready.

yaswant commented on .github/workflows/build.yml:
```diff
@@ -71,6 +71,12 @@ jobs:
+      - name: Set up git configuration
+        run: |
+          git config --global user.email "fab"
+          git config --global user.name "Fab Testing"
```
I see you need this for `git merge`. There is no potential issue here, but does this have to be --global config? I know there is no strict validation here, but wondering if fab@github.action as user.email would be better?

t00sa replied:
It probably doesn't have to be, but this is what git suggests when the configuration items don't exist

t00sa commented:
I have addressed the comments from the previous reviews, the CI tests pass, and I've successfully used my branch to build skeleton in LFRic core using Jason's branch, so I think this ready for another round of reviewing.

hiker commented:
Removed 'ready for review' label, since it's still a draft.
